docs/proposals/ktl.md

```diff
@@ -0,0 +1,243 @@
+
+And also:
+
+- allow the cli to be released independently from the cluster
+- allow sub command groups within the cli to be released independently from one another
+- allow a decentralized ecosystem of tools to leverage centralized maintained cli / client infrastructure
+- facilitate end-to-end ownership of the cli, and in some cases sub command groups with the cli
+- facilitate decentralized development of extensions for the cli
+
+## Anti-goals
+
+Things we want to avoid.
+
+- block on moving existing kubectl commands out of the kubernetes/kubernetes repo
+- rewrite kubectl from the ground up in a new repo
+
+## Non-goals
+
+Even if these are good ideas, don't let them distract us from meeting our goals will simpler solutions.
+
+- build solution for discovering installable plugins and installing them
+  - rely on existing package management solutions for this until we need something more
+- invent new build and distribution infrastructure
+- fix issues with the existing kubectl commands
+- dog fooding the plugin mechanism for core commands
+
+## Reference-level explanation
+
+- build a new cli binary `ktl` (kubectl2) under the `kubernetes/ktl` repo that dispatches to commands developed in other repos
+- keep old cli commands in `kubernetes/kubernetes/cmd/kubectl` and vendor them into `kubectl/ktl`
+- build new cli commands in `kubernetes/kubectl` and vendor them into `kurnetes/ktl`
+- build common cli Kubernetes client infrastructure and libraries that can be used to develop a decentralized cli ecosystem
+
+### ktl
+
+In a new repo (`kubernetes/ktl`), create the `ktl` binary that dispatches (both statically and dynamically)
+to commands developed in other repos.
+
+#### Dispatch
+
+Static dispatch:
+
+- vendor in kubernetes/kubernetes/cmd/kubectl cobra commands, and register them under `ktl kube`
+- vendor in kubernetes/kubectl/cmd cobra commands, and register them under `ktl`
+- vendor in commands from other sources as needed over time
+
+Dynamic dispatch:
+
+- support git-style plugins that allow registering new commands under `ktl`
+  - use simplified version of the kubectl plugin implementation - make configuration files optional
+- plugins only purpose is discovery of kubernetes related commands
+  - plugins can leverage shared client cli libraries (whether they are installed as plugins or not)
+- by default throw an error if plugin names conflict with existing commands
+  - this is configurable
+- plugins can be disabled through modifying `~/.ktlconfig`
+
+Overriding existing commands:
+
+- support command alias' to allow overriding one command with another
+  - allows plugins to extend and override built-in commands
+  - allows `ktl kube *` commands to be aliased to `ktl *`
+
+#### Configuration
+
+Use [viper](https://github.com/spf13/viper) to configure dispatch precedence
+
+### Cli commands
+
+All new cli commands should be built outside of kubernetes/kubernetes and vendored into the kubernetes/ktl.  Command repos
```

Comment:
Can you explain the benefit of forcing a layer of vendoring?  I see a distinction between developing re-useable CLI infrastructure (if that's a thing we thing we want to do) and writing commands.  By all means, vendor the infrastructure, but why vendor the commands too?

Reply:
1) It keeps kubernetes/kubernetes outside of kubectl
2) Each vendored set of commands can have a separate release cycle from the others

Reply:
> It keeps kubernetes/kubernetes outside of kubectl
> Each vendored set of commands can have a separate release cycle from the others

Picking winners and losers like this along with vendoring concerns seems contrary to goals of being generic and having a simple core.  If we do this, we could instead focus on the simple core.  We have a client (`kubectl`) that already works.  We should be careful to avoid compromises in a new tool to have quick progress we'll later regret.

Reply:
In my experience rapid iteration and continual refactoring + adjustment over time typically will have better results than careful planning with slow iteration cycles.

Reply:
Note that by adding a new CLI tool `ktl` we are imposing an enormous cost on the kubernetes ecosystem and users.  That's not free.

Reply:
Is this because of fragmentation?